### DISCUSSION TITLE
Backend OpenAPI specification

### DISCUSSION THREAD

**Discussion by chriskapp** [eclipse-theia/theia#7784] [General]
Hi, Iam just curious is there an OpenAPI specification which describes the basic API endpoints, which can be used i.e. to implement the server side of the IDE in another language instead of NodeJS?

**Comment by vince-fugnitto** [comment]
@chriskapp there is nothing at the moment, the following issue might provide more information for you: https://github.com/eclipse-theia/theia/issues/3396

**Comment by chriskapp** [comment]
Thanks for the information. I have found an API documentation about Eclipse-Orion https://wiki.eclipse.org/Orion/Server_API which is used by Eclipse Che, Iam not really sure about the relationship to Eclipse Theia but is this probably the same Backend API?

**Reply by vince-fugnitto** [reply]
No, they are completely different and separate projects so they do not share the same backend API. Eclipse Theia is a framework to build IDE-like applications and Che is among one of many applications which use the framework. For additional information regarding the relationship between Che and Theia you can consult the following [article](https://eclipsesource.com/blogs/2018/12/03/eclipse-che-vs-eclipse-theia/)  from EclipseSource. I hope this helps clarify things a little bit better :)

**Reply by chriskapp** [reply]
ok, thanks for clarifying this 👍

**Reply by gorkem** [reply]
Also Che does not use Orion Server API.

**Comment by akosyakov** [comment]
Backend APIs are JSON-RPC protocols wrapped in statically typed Typescript interfaces, look for files named `*-protocol.ts`. Usually they belong to `common` module of the extension package.

Are you looking for particular? What for?

**Comment by chriskapp** [comment]
So we are currently working on an API management system where we need the capabilities to edit files via the web. For this we would like to use a full fledged editor like Theia. We need basic file editing, git and a shell. But your hint is already very valuable and it is great that Theia uses JSON-RPC. Is this served over web-sockets or does this work with plain HTTP, since in our case we are limited to HTTP.

**Reply by akosyakov** [reply]
JSON-RPC endpoints are bi-directional, so HTTP won't do for them. All JSON-RPC endpoints are multiplexed over the same WS. There are some HTTP endpoints for heavy stuff like file download, since WS should not be blocked by big messages. Theia extension can configure own HTTP endpoints by implementing `BackendApplicationContribution.configure(app)`.

I don't think you can implement backend in another runtime or replace WS transport. In theory it is possible, in practice it will be maintenance pain. I would either try to wrap up your HTTP endpoints in some frontend concepts, like `Resource`, or JSON-RPC endpoint and then use it on the frontend.
